---
title: "The effect of sequencing depth on the performance of genotyper"
author: "Zexuan Zhao"
date: "2023-02-26"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(ggpubr)
library(scales)
library(tidyverse)
```

Workflow:

1. Sampling reads with 5x or 10x depth.

2. Mapping reads to W2A reference genome.

3. Joint genotyping. Reads with the same depth and from the same species are called together. 

4. Filter variants that are in TE elements.

4. Compare genotypes generated from 5x or 10x read depth to those generated by all reads for each sample. The effect of TE filtering is shown by comparing the metrics calcualted before and afther TE filtering. 


Definition of confusion matrix in the context of genotyping:

- true-positives (TP) : variants/genotypes that match in truth (all reads) and query (5x or 10x).

- false-positives (FP) : variants that have mismatching genotypes or alt alleles, as well as query variant calls in regions a truth set would call confident hom-ref regions.

- false-negatives (FN) : variants present in the truth set, but missed in the query.


## Overall genotyping precision and recall

This table represents the results in ratios:

- Recall:	Recall for truth variant representation = TRUTH.TP / (TRUTH.TP + TRUTH.FN). Recall represents the sensitivity of the genotyper. Low recall means many variants are missed.
- Precision:	Precision of query variants = QUERY.TP / (QUERY.TP + QUERY.FP). Precision represents the specificity of the genotyper. Low precision means many variants called are spurious.
- F1_Score: 2 * Precision * Recall / (Precision + Recall). This metrics summarize the overall performance (sensitifity and specifisity).


```{r}
tb <- read_tsv("~/Downloads/QC by sampling - QC by sampling.tsv")
```


```{r}
p1 <- tb %>% 
  group_by(Depth, TE_filtered) %>% 
  summarize(mean_Recall = mean(Recall), sd_Recall = sd(Recall)) %>% 
  ungroup() %>% 
  mutate(Depth = as.factor(Depth)) %>% 
  mutate(TE_filtered = ifelse(TE_filtered, "Non-TE", "All")) %>% 
  ggplot(aes(x=Depth, y=mean_Recall, fill = TE_filtered)) +
    geom_col(position = position_dodge(width=.9)) +
    geom_errorbar(aes(x=Depth, ymin=mean_Recall-sd_Recall, ymax=mean_Recall+sd_Recall),
                  position = position_dodge(width=.9),
                  width=0.4) +
  scale_y_continuous(limits=c(0.9,1),oob = rescale_none) +
  ylab("Recall") +
  xlab("Depth") +
  guides(fill = guide_legend(title="Variants")) +
  theme_classic()
```

```{r}
p2 <- tb %>% 
  group_by(Depth, TE_filtered) %>% 
  summarize(mean_Precision = mean(Precision), sd_Precision = sd(Precision)) %>% 
  ungroup() %>% 
  mutate(Depth = as.factor(Depth)) %>% 
  mutate(TE_filtered = ifelse(TE_filtered, "Non-TE", "All")) %>% 
  ggplot(aes(x=Depth, y=mean_Precision, fill = TE_filtered)) +
    geom_col(position = position_dodge(width=.9)) +
    geom_errorbar(aes(x=Depth, ymin=mean_Precision-sd_Precision, ymax=mean_Precision+sd_Precision),
                  position = position_dodge(width=.9),
                  width=0.4) +
  scale_y_continuous(limits=c(0.9,1),oob = rescale_none) +
  ylab("Precision") +
  xlab("Depth") +
  guides(fill = guide_legend(title="Variants")) +
  theme_classic()
```

```{r}
p3 <- tb %>% 
  group_by(Depth, TE_filtered) %>% 
  summarize(mean_F1_Score = mean(F1_Score), sd_F1_Score = sd(F1_Score)) %>% 
  ungroup() %>% 
  mutate(Depth = as.factor(Depth)) %>% 
  mutate(TE_filtered = ifelse(TE_filtered, "Non-TE", "All")) %>% 
  ggplot(aes(x=Depth, y=mean_F1_Score, fill = TE_filtered)) +
    geom_col(position = position_dodge(width=.9)) +
    geom_errorbar(aes(x=Depth, ymin=mean_F1_Score-sd_F1_Score, ymax=mean_F1_Score+sd_F1_Score),
                  position = position_dodge(width=.9),
                  width=0.4) +
  scale_y_continuous(limits=c(0.9,1),oob = rescale_none) +
  ylab("F1 Score") +
  xlab("Depth") +
  guides(fill = guide_legend(title="Variants")) +
  theme_classic()
```

```{r}
p <- ggarrange(p1, p2, p3,
          ncol = 1, nrow = 3,
          common.legend = TRUE,
          legend = "bottom")
ggsave("../figures/QC by sampling.png", plot = p,
        width = 210, height = 360, units = "mm")
```