---
title: "Calculate TE density"
author: "Zexuan Zhao"
date: "2023-06-06"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(biomartr)
library(GenomicRanges)
library(Biostrings)
library(parallel)
library(tidyverse)

# Input size
TE_annotation_file <- "../data/v4.W2A.TE_annotation.TEanno.gff3"
genome_file <- "../data/v4.1.W2A_genome_assembly.fasta"
fai_file <- "../data/v4.1.W2A_genome_assembly.fasta.fai"
out_file <- "W2A_TE_density_by_superfamilies.tsv"

# Window size
win_size <- 200 * 1000 
```

Color scheme

```{r}
colors_simplified <- c("Retrotransposon/LTR/Copia" = "#f9584b",
                       "Retrotransposon/LTR/Gypsy" = "#ff8980",
                       "DNA transposons/Helitron/Helitron" = "#CD7F32",
                       "DNA transposons/TIR/Tc1_Mariner" = "#CC5500",
                       "DNA transposons/TIR/Mutator" = "#FFAC1C",
                       "DNA transposons/TIR/CACTA" = "#ffe173",
                       "Other superfamilies" = "#80e8e0",
                       "Unclassified TE" = "#ccd6db")
```

Utils

```{r}
chr_lengths <- read_tsv(fai_file,
                        col_names = c("chr","length","bits","spacer1","spacer2")) %>% 
  dplyr::select(chr, length) %>% 
  filter(str_detect(chr, "chr"))


short_to_long_name <- tibble(superfamily = c("Copia", "Gypsy", "Helitron", "Tc1_Mariner", "Mutator", "CACTA", "other_superfamilies", "Unclassified"),
                             superfamily_long = names(colors_simplified))
```


## 1. Read in annotation and data cleaning

```{r}
# Read in gff file
TE_annotation <- read_gff(TE_annotation_file) %>% 
  filter(str_detect(attribute, "Method=homology")) %>% 
  mutate(strand = "+") %>% # Put all TE on + strand, so that overlapping TEs can be merged
  filter(str_detect(seqid, "chr")) # Filter non-chromosomal TEs
# Convert to gr format
TE_annotation_gr <- makeGRangesFromDataFrame(TE_annotation) %>% 
  GenomicRanges::reduce()# Merge overlapping TE regions
```


```{r}
# separate super families
Copia <- TE_annotation %>% 
  filter(type == "Copia_LTR_retrotransposon") %>% 
  mutate(type = "Retrotransposon/LTR/Copia")
Gypsy <- TE_annotation %>% 
  filter(type == "Gypsy_LTR_retrotransposon") %>% 
  mutate(type == "Retrotransposon/LTR/Gypsy")
Helitron <- TE_annotation %>% 
  filter(type == "helitron") %>% 
  mutate(type == "DNA transposons/Helitron/Helitron")
Tc1_Mariner <- TE_annotation %>% 
  filter(type == "Tc1_Mariner_TIR_transposon") %>% 
  mutate(type == "DNA transposons/TIR/Tc1_Mariner")
Mutator <- TE_annotation %>% 
  filter(type == "Mutator_TIR_transposon") %>% 
  mutate(type == "DNA transposons/TIR/Mutator")
CACTA <- TE_annotation %>% 
  filter(type == "CACTA_TIR_transposon") %>% 
  mutate(type == "DNA transposons/TIR/CACTA")
other_superfamilies <- TE_annotation %>% 
  filter(type %in% c("PIF_Harbinger_TIR_transposon", "hAT_TIR_transposon", "Penelope_retrotransposon")) %>% 
  mutate(type == "Other superfamilies")
unclassified <- TE_annotation %>% 
  filter(type %in% c("repeat_region", "LTR_retrotransposon", "LINE_element", "DNA_transposon")) %>% 
  mutate(type == "Unclassified TE")
```


```{r}
# Convert to gr format
Copia_gr <- Copia %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce()
Gypsy_gr <- Gypsy %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce()
Helitron_gr <- Helitron %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce()
Tc1_Mariner_gr <- Tc1_Mariner %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce()
Mutator_gr <- Mutator %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce()
CACTA_gr <- CACTA %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce()
other_superfamilies_gr <- other_superfamilies %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce()
unclassified_gr <- unclassified %>% 
  makeGRangesFromDataFrame() %>% 
  GenomicRanges::reduce()
```

## 2. Add genome info to annotation

```{r}
genome <- readDNAStringSet(genome_file)
genome_info <- GenomeInfoDb::Seqinfo(seqnames = names(genome)[str_detect(names(genome), "chr")],
                                     seqlengths = width(genome)[str_detect(names(genome), "chr")],
                                     isCircular = NA,
                                     genome = NA)
seqinfo(Copia_gr) <- genome_info
seqinfo(Gypsy_gr) <- genome_info
seqinfo(Helitron_gr) <- genome_info
seqinfo(Tc1_Mariner_gr) <- genome_info
seqinfo(Mutator_gr) <- genome_info
seqinfo(CACTA_gr) <- genome_info
seqinfo(other_superfamilies_gr) <- genome_info
seqinfo(unclassified_gr) <- genome_info
```

## 3. Calculate TE density

Define function

```{r}
calculate_density <- function(gr, win_size) {
  # Singular function for mclapply
  # Calculate proportion of bps in window overlapped with gr
  # Return a data frame
  cal_overlap <- function(window, gr){
    overlap <- GenomicRanges::intersect(gr, window)
    n_overlaps <- overlap %>% width() %>% sum()
    return(tibble(chr = as.character(window@seqnames),
                  pos = mid(window),
                  n = n_overlaps))
  }
  
  # 1. Make windows
  # The last window of each chromosome is dropped due to insufficient size
  windows <- tileGenome(gr@seqinfo, 
                        tilewidth = win_size, 
                        cut.last.tile.in.chrom = TRUE)
  strand(windows) <- "+"
  windows <- windows[width(windows) == win_size]
  
  # Run cal_overlap
  tb <- mclapply(windows%>% split(), cal_overlap, gr = gr, mc.cores = 8)
  # Merge result and return
  result <- do.call(bind_rows, tb)
  result
}
```

Run:

```{r eval = TRUE}
## Call calculate_proportions
Copia_density <- calculate_density(Copia_gr, win_size) %>% 
  mutate(p = n/win_size)
Gypsy_density <- calculate_density(Gypsy_gr, win_size) %>% 
  mutate(p = n/win_size)
Helitron_density <- calculate_density(Helitron_gr, win_size) %>% 
  mutate(p = n/win_size)
Tc1_Mariner_density <- calculate_density(Tc1_Mariner_gr, win_size) %>% 
  mutate(p = n/win_size)
Mutator_density <- calculate_density(Mutator_gr, win_size) %>% 
  mutate(p = n/win_size)
CACTA_density <- calculate_density(CACTA_gr, win_size) %>% 
  mutate(p = n/win_size)
other_superfamilies_density <- calculate_density(other_superfamilies_gr, win_size) %>% 
  mutate(p = n/win_size)
unclassified_density <- calculate_density(unclassified_gr, win_size) %>% 
  mutate(p = n/win_size)
```

# 4. Plot

```{r}
pos2cumulative_pos <- function(chromosome, position, chr_lengths){
  chr_lengths_cumulative <- chr_lengths %>% 
    mutate(total=cumsum(length)-length) %>%
    dplyr::select(-length)
  tibble(chr = chromosome,
         pos = position) %>% 
    left_join(chr_lengths_cumulative) %>% 
    mutate(cumulative_pos = pos + total) %>% 
    pull(cumulative_pos)
}

plot_along_genome_scaffold <- function(data, chr_lengths){
  axisdf <- chr_lengths %>% 
    mutate(total = cumsum(length) - length) %>%
    mutate(center = total + length/2)
  data %>% 
    ggplot(aes(x = cumulative_pos)) +
      scale_color_manual(values = rep(c("grey", "skyblue"), nrow(chr_lengths))) +
      scale_x_continuous(label = axisdf$chr, breaks= axisdf$center) +
      theme_bw() +
      theme(legend.position="none",
            panel.border = element_blank(),
            panel.grid.major.x = element_blank(),
            panel.grid.minor.x = element_blank())
}
```

```{r}
p_Copia <- Copia_density %>% 
  mutate(cumulative_pos = pos2cumulative_pos(chr, pos, chr_lengths)) %>% 
  plot_along_genome_scaffold(chr_lengths = chr_lengths) +
    geom_point(aes(y = p, color = as.factor(chr),  alpha = 0.8), size=1.3) +
    ylab("Copia density (%)") +
    xlab("Chromosome")
p_Gypsy <- Gypsy_density %>% 
  mutate(cumulative_pos = pos2cumulative_pos(chr, pos, chr_lengths)) %>% 
  plot_along_genome_scaffold(chr_lengths = chr_lengths) +
    geom_point(aes(y = p, color = as.factor(chr),  alpha = 0.8), size=1.3) +
    ylab("Gypsy density (%)") +
    xlab("Chromosome")
p_Helitron <- Helitron_density %>% 
  mutate(cumulative_pos = pos2cumulative_pos(chr, pos, chr_lengths)) %>% 
  plot_along_genome_scaffold(chr_lengths = chr_lengths) +
    geom_point(aes(y = p, color = as.factor(chr),  alpha = 0.8), size=1.3) +
    ylab("Helitron density (%)") +
    xlab("Chromosome")
p_Tc1_Mariner <- Tc1_Mariner_density %>% 
  mutate(cumulative_pos = pos2cumulative_pos(chr, pos, chr_lengths)) %>% 
  plot_along_genome_scaffold(chr_lengths = chr_lengths) +
    geom_point(aes(y = p, color = as.factor(chr),  alpha = 0.8), size=1.3) +
    ylab("Tc1 Mariner density (%)") +
    xlab("Chromosome")
p_Mutator <- Mutator_density %>% 
  mutate(cumulative_pos = pos2cumulative_pos(chr, pos, chr_lengths)) %>% 
  plot_along_genome_scaffold(chr_lengths = chr_lengths) +
    geom_point(aes(y = p, color = as.factor(chr),  alpha = 0.8), size=1.3) +
    ylab("Mutator density (%)") +
    xlab("Chromosome")
p_CACTA <- CACTA_density %>% 
  mutate(cumulative_pos = pos2cumulative_pos(chr, pos, chr_lengths)) %>% 
  plot_along_genome_scaffold(chr_lengths = chr_lengths) +
    geom_point(aes(y = p, color = as.factor(chr),  alpha = 0.8), size=1.3) +
    ylab("CaCTA density (%)") +
    xlab("Chromosome")
p_other_superfamilies <- other_superfamilies_density %>% 
  mutate(cumulative_pos = pos2cumulative_pos(chr, pos, chr_lengths)) %>% 
  plot_along_genome_scaffold(chr_lengths = chr_lengths) +
    geom_point(aes(y = p, color = as.factor(chr),  alpha = 0.8), size=1.3) +
    ylab("Other superfamily density (%)") +
    xlab("Chromosome")
p_unclassified <- unclassified_density %>% 
  mutate(cumulative_pos = pos2cumulative_pos(chr, pos, chr_lengths)) %>% 
  plot_along_genome_scaffold(chr_lengths = chr_lengths) +
    geom_point(aes(y = p, color = as.factor(chr),  alpha = 0.8), size=1.3) +
    ylab("Unclassified TE density (%)") +
    xlab("Chromosome")
```

```{r}
ggpubr::ggarrange(p_Copia, p_Gypsy, p_Helitron, p_Tc1_Mariner, p_Mutator,p_CACTA, p_other_superfamilies,p_unclassified,
                  ncol = 1)

ggsave("../figures/TE_density_by_superfamilies(separated).png", width = 210, height = 297, units = "mm")
```

## Plot2

```{r}
axisdf <- chr_lengths %>% 
    mutate(total = cumsum(length) - length) %>%
    mutate(center = total + length/2)

Copia_density %>% mutate(superfamily  = "Retrotransposon/LTR/Copia") %>% 
  bind_rows(Gypsy_density %>% mutate(superfamily  = "Retrotransposon/LTR/Gypsy")) %>% 
  bind_rows(Helitron_density %>% mutate(superfamily  = "DNA transposons/Helitron/Helitron")) %>% 
  bind_rows(Tc1_Mariner_density %>% mutate(superfamily  = "DNA transposons/TIR/Tc1_Mariner")) %>% 
  bind_rows(Mutator_density %>% mutate(superfamily  = "DNA transposons/TIR/Mutator")) %>% 
  bind_rows(CACTA_density %>% mutate(superfamily  = "DNA transposons/TIR/CACTA")) %>% 
  bind_rows(other_superfamilies_density %>% mutate(superfamily  = "Other superfamilies")) %>% 
  bind_rows(unclassified_density %>% mutate(superfamily  = "Unclassified TE")) %>% 
  left_join(short_to_long_name) %>% 
  mutate(cumulative_pos = pos2cumulative_pos(chr, pos, chr_lengths)) %>% 
  ggplot(aes(x= cumulative_pos, y = p*100, col = superfamily)) +
    tidyquant::geom_ma() +
    #geom_point(alpha = 0.1, size = 0.1) +
    scale_x_continuous(label = axisdf$chr, breaks= axisdf$center) +
    theme_classic() +
    scale_color_manual(values = colors_simplified)+
    guides(color=guide_legend(title="TE superfamilies")) +
    ylab("TE Density (%)")+
    xlab("Chromosome")

ggsave("../figures/TE_density_by_superfamilies(merged).png", width = 210, height = 120, units = "mm")
```

